go/logic/applier.go

```diff
@@ -1043,3 +1062,73 @@ func (this *Applier) Teardown() {
 	this.singletonDB.Close()
 	atomic.StoreInt64(&this.finishedMigrating, 1)
 }
+
+func (this *Applier) SaveExistingGTIDs() error {
+	query := `select /* gh-ost */ @@global.server_uuid as server_uuid, @@global.gtid_executed as gtid_executed`
+	err := sqlutils.QueryRowsMap(this.singletonDB, query, func(rowMap sqlutils.RowMap) error {
+		this.pristineGTIDSet = NewGTIDSet(
+			rowMap.GetString("server_uuid"),
+			rowMap.GetString("gtid_executed"),
+		)
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (this *Applier) PurgeGTIDs() error {
+	if _, err := sqlutils.ExecNoPrepare(this.singletonDB, `PURGE BINARY LOGS BEFORE NOW()`); err != nil {
```

Comment:
Note I don't feel strongly about this `PURGE BINARY LOGS BEFORE NOW()`

This may change over to `PURGE BINARY LOGS TO ...`, I just didn't understand @evanelias's warning on https://github.com/github/gh-ost/issues/254#issuecomment-502524514

However per @shlomi-noach's comment on https://github.com/github/gh-ost/issues/254#issuecomment-504996069 I'm not sure if we even need to run any purge. Going to wait until the discussion ends first before changing anything.

Reply:
My concern is there may be some cases where `PURGE BINARY LOGS BEFORE NOW()`, when run after the GTID surgery, might not actually purge the first and only binlog. `NOW()` does not have sub-second precision. So if that binlog was created in the current clock second, does it actually get purged? In other words, is `BEFORE` like `<` or is it like `<=`?

If that is indeed a problem, it may be avoidable with something goofy like `NOW() + INTERVAL 1 DAY`. But using `SHOW MASTER STATUS` to get the current binlog feels more precise, in my opinion.

Reply:
> NOW() does not have sub-second precision

Ahh thanks! Will change.